### PR TITLE
36038 fix selection list footer extra padding

### PR DESCRIPTION
**Pull request by lukemorawski**
### Details

This PR fixes extra padding below Referral banner in Search list.
The issue was due to the fact that the safe area bottom padding was added two times - one by `SelectionList` and second by `ScreenWrapper`. After further investigation it turned out that this problem occurs in many other pages/components. In most cases it isn't as much as pronounced as on the SearchPage, as it's just an extra large space on the bottom of the list.
Every screen that uses `SelectionList` have been inspected and corrected if necessary.
A second element of this issue was when `SearchList` had a `footerContent` passed as params but that component was returning null or an empty element. `SearchList` would still create a container for that footer, with it's own margins and padding, taking up space at the bottom of the list even though, nothing was displayed there.
This issue was taken care of by checking the layout dimensions of the `children` passed to the `SelectionList` component and removing the footer from the layout if it's children have nothing to render. Two new hooks are taking care of this `useElementLayoutVisibility` and `useThrottledState`. 

### Fixed Issues

$ https://github.com/Expensify/App/issues/36038
PROPOSAL: no proposal

### Tests

- same as QA
- [ ] Verify that no errors appear in the JS console

### Offline tests

same as QA

### QA Steps

- [ ] open app on IOS and tap search
- [ ] observe the referral banner at the bottom
Other places where changes where introduced:
- [ ] Request Money Participant Selection Step
- [ ] Send Money Participant Selection Step
- [ ] Money Request Currency Selector
- [ ] Workspace Invite Page
- [ ] Room Invite Page
- [ ] Room Members Page
- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/134388952/c4bc9107-fd5a-4b3a-86af-6e1782cc4101

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/134388952/4007c143-972d-455c-a63b-9020c4a1442d

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/134388952/e36314d5-ea8a-4c3c-9b68-c625b1c29274

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/134388952/6cccfc0d-0347-41b9-8021-cb1a77e24bef

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/134388952/e560122e-3006-49fa-86e0-331316fb19a3

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/134388952/6c7b42ec-c721-4b3e-bc78-1f46adde9ba8

</details>

### CODE REVIEW

File: src/hooks/useLayoutVisibility.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            const {height, width} = event.nativeEvent.layout;
+            const visible = height > 0 && width > 0;
+
+            // Use the throttled version of setIsVisible
+            setIsVisible(visible);
+        },
+        [setIsVisible],
+    );
+
+    return [onLayout, isVisible];
```
[the remark this answers is not in the document]
Comment (lukemorawski) [reply src/hooks/useLayoutVisibility.ts:35]: It was actually switched places on purpose. You can't use the flag without setting the layout change event handler. Well, you can but you won't get much out of it :) So to ensure and encourage setting the handler I've put it as the first value of the tuple.

[the remark this answers is not in the document]
Comment (rezkiy37) [reply src/hooks/useLayoutVisibility.ts:35]: Hm... I got you idea. It is not a critical point. So, if you think this way, let's keep it as it is.

**Review by luacmartins — CHANGES_REQUESTED** [review]
@lukemorawski we have conflicts now

**Comment by abdulrahuman5196** [comment]
@rezkiy37 

>A second element of this issue was when SearchList had a footerContent passed as params but that component was returning null or an empty element. SearchList would still create a container for that footer, with it's own margins and padding, taking up space at the bottom of the list even though, nothing was displayed there.
This issue was taken care of by checking the layout dimensions of the children passed to the SelectionList component and removing the footer from the layout if it's children have nothing to render. Two new hooks are taking care of this useElementLayoutVisibility and useThrottledState.

This seems to be complicated fix for a simpler issue.
Why can't clients of selection list take responsibility of sending `null` value if no footer is available to show?
This way selection list can, just not add the footer containers if footer doesn't have value.

Do let me know if I am missing something here.

**Comment by rezkiy37** [comment]
> @rezkiy37
> 
> > A second element of this issue was when SearchList had a footerContent passed as params but that component was returning null or an empty element. SearchList would still create a container for that footer, with it's own margins and padding, taking up space at the bottom of the list even though, nothing was displayed there.
> > This issue was taken care of by checking the layout dimensions of the children passed to the SelectionList component and removing the footer from the layout if it's children have nothing to render. Two new hooks are taking care of this useElementLayoutVisibility and useThrottledState.
> 
> This seems to be complicated fix for a simpler issue. Why can't clients of selection list take responsibility of sending `null` value if no footer is available to show? This way selection list can, just not add the footer containers if footer doesn't have value.
> 
> Do let me know if I am missing something here.

@abdulrahuman5196, sorry but it is not my PR. @lukemorawski is an author.

**Comment by lukemorawski** [comment]
@abdulrahuman5196 I believe your question was addressed to me?
There are cases where the client component where `SelectionList` is instantiated doesn't have the control over what is in the footer. This is when a external component is passed as footer content, such as `ReferralProgramCTA`. Moving that logic to the client component would severely impact DRY principle. That way we can have a one place where a referral program banner can be displayed and dismissed without implementing this in every every component where an instance of `SelectionList` is used. But obviously I'm open for suggestions

**Comment by lukemorawski** [comment]
@abdulrahuman5196 yes, but the you need implement this conditional everywhere `{!dismissedReferralBanners[referralContentType] && ...` instead of just one place which is the `ReferralProgramCTA`. My approach allows to have the dismissal logic in just one place. Or say you have a footer component, that conditionally renders 3 components, but wrapped in some styled View element, and all of those components are not rendered (because of falsy conditions). This aproach ensures that there won't be any empty space in the list footer.

**Comment by abdulrahuman5196** [comment]
@lukemorawski 

>Or say you have a footer component, that conditionally renders 3 components, but wrapped in some styled View element, and all of those components are not rendered (because of falsy conditions)

Even in this case, I think the responsibility should be with the footer component to not render anything. or as our case send null to parent component.
We shouldn't delegate the responsibility of the child/client component to parent. If we think of this case, then every core component would be subjectable to this issue, since any child/client component can pass up `empty view` to parent/core component, and we would end up adding same logic to all core components. 
So I am still not aligned with checking the visibility of child/client components to solve the empty view issue.


>but the you need implement this conditional everywhere {!dismissedReferralBanners[referralContentType] && ... instead of just one place which is the ReferralProgramCTA

If code duplication is the problem, we should think of having common helpers or something else to solve the code duplication. But not to mix/change the responsibility.

@luacmartins What do you think here?

**Comment by luacmartins** [comment]
>we should think of having common helpers or something else to solve the code duplication. But not to mix/change the responsibility.

I agree with this.

**Comment by lukemorawski** [comment]
@luacmartins @abdulrahuman5196 ok, i'll try to come up with some other solution.

**Pull request closed without merging**